spmd/tensor/parallel/api.py

```diff
@@ -76,12 +76,75 @@ def _shard_self_attn_params(name: str, module: nn.Module) -> None:
                 module.register_module(n, tp_multi_head_attention)
 
 
+def _is_mlp(module: nn.Module) -> bool:
```

Comment:
The logic here looks good to me. Actually after a second thought. I think let's do something like this:

1. _get_linear_module_count():
2. For Pairwise parallel style, we will throw errors if the count == 0 or the count % 2 != 0.

Because at the end of day, MLP literally means multiple layers of Linear. And we just need to ensure that it's flattened and error out if it is a nested version. WDT?

Reply:
What do you mean by "a nested version"? For simplicity, we assume that the low-level API _parallelize_mlp() is only called on a bottom-level MLP sub-module where we simply check if the parity of linear layers' count matches with the parallel style being passed.

Reply:
Right, but I guess what I am trying to say is that:
1. Nested I mean the module is not flat, like Sequence(Linear(), LayerNorm(Linear()), Linear()) something like this. Because children only returns all direct children and it's not recursive.
2. We should not assume MLP only has even-number of layers. Instead, this constraint should go with pairwise parallel style.